Assets/Src/Scripts/AgileMvvm/ViewModel.cs

```diff
@@ -0,0 +1,143 @@
+
+namespace AgileMvvm {
+  // The ViewModel class that maintains models and raises model update events to the bound Views. It
+  // also provides the utility methods to manage the binding relationships.
+  //
+  // In AgileMvvm, the relationships among View, ViewModel and Model are:
+  //
+  // * The Model does not know about the View or the ViewModel.
+  // * The View does not know about the Model, but knows the ViewModel.
+  // * The View may own the ViewModel, but it is not necessary to do that. The View can communicate
+  //   with the ViewModel via lightweight bindings even when it doesn't own the ViewModel.
+  // * The binding relationships between the View and the ViewModel can be set up either by the View
+  //   itself, or by the Application class, the GameManager class, etc.
+  // * Different Views or multiple View instances can be bound to the same ViewModel.
+  // * The ViewModel does not know about the View, but knows and maintains the Model.
+  //
+  // Typical MVVM patterns support the following binding modes between the View and the ViewModel:
+  //
+  // * One-way binding: ViewModel's property to View's property or View's event handler.
+  // * One-way-to-source binding: View's property to ViewModel's property. (Rarely used)
+  // * Two-way binding: Two-way update between View's property and ViewModel's property.
+  // * Event binding: View's event to ViewModel's event handler.
+  //
+  // As an lightweight framework, AgileMvvm explicitly supports the one-way binding mode via the
+  // ViewModel class. To establish a one-way binding between the View and the ViewModel:
+  //
+  // * Defines your own ViewModel class and implement bindable properties like below:
+  //
+  //    class Person : ViewModel {
+  //      private string _name;
+  //      [BindableProperty]
+  //      public string Name {
+  //        get => _name;
+  //        set => MvvmSetter(ref _s, value);
+  //      }
+  //    }
+  //
+  // * Defines your View class:
+  //
+  //    class Dialog {
+  //      public string TextField { get; set; }
+  //    }
+  //
+  // * Binds the bindable property of the ViewModel to the target property of the View class:
+  //
+  //    var dialog = new Dialog();
+  //    var person = new Person();
+  //    person.Bind("Name", dialog, "TextField");
+  //
+  // That's it! Any subsequent changes of the ViewModel's bound property will update the View's
+  // property.
+  //
+  // * Or, you can define your own event handler if you want to customize the binding behavior:
+  //
+  //    var dialog = new Dialog();
+  //    var person = new Person();
+  //    person.Bind("Name", (object sender, UpdatedEvent.Args e) => {
+  //      dialog.TextField = e.Value.ToString();
+  //    });
+  //
+  // As for other binding modes, AgileMvvm doesn't provide direct support so far. But it is easy to
+  // implement them with C#'s event handling mechanism on top of AgileMvvm.
+  //
+  // TODO: Consider to support other binding modes directly.
```

Comment:
Your implementation is quit similar with the [Microsoft.Toolkit.Mvvm package](https://docs.microsoft.com/en-us/windows/communitytoolkit/mvvm/introduction). Have you considered to use this package directly?

Reply:
First, I don't like the interface design of `Microsoft.Toolkit.Mvvm` personally. It has many complex and academic naming conventions, such as `ObservableRecipient`, `ISettingsService`, etc. It has the burden to support WPF-like bindings via XAML format.

Second, Unity has so many differences compared to a WPF application. It's way easier to start from a super lightweight framework than to follow an existing design.

Third, I'd avoid introducing new DLL/NuGet dependencies to the Unity project.

BTW, my implementation is very like a tiny subset of https://www.mvvmcross.com/, another C# MVVM framework that is heavier than `Microsoft.Toolkit.Mvvm`.